### PR TITLE
machines: Support for external providers

### PR DESCRIPTION
**Pull request by mareklibra**
With this patch, data and actions for the 'machines' package can be
provided by pluggable external providers.

Their use is optional. If none is installed, fails to load or to initialize,
the Libvirt is used by default.

This patch allows reuse of common UI for different virtualization providers while
keeping upstream independent on them. An external providers bring more realistic view
of a virtualization host if the server is used in single-purpose installation.

The providers can be chained - i.e. certain operation can be still performed
via Libvirt instead of external system.

Provider-specific code moved to the Libvirt provider (i.e. VM state manipulation).

The external providers are supposed to be implemented separatly from the cockpit-project code base.

A provider is installed by copy of it's index.js file under /usr/share/cockpit/machines/provider and loaded dynamically.

For provider's API and further details, please refer the machines/README.md .

### CODE REVIEW

**Comment by stefwalter** [comment]
Hey @mareklibra nice work.

This is not a blocker, but I noticed your use case here and wanted to see if i could cover it with some work that I'm doing for translations: I'd like to add a line to manifest.json which describes how a certain filename in a package is concatenated together from other filenames. I'm going to be using a glob style wildcard.

So for example. If I had this in my manifest.json:

```
"resources": {
   "provider.js": "provider-*.js"
}
```

Then when ```/@localhost/package/provider.js``` is requested ... cockpit-bridge will serve the concatenation of all files in the package that match ```provider-*.js``` ... so for example ```provider-libvirt.js``` ```provider-machined.js``` and so on. The order will be alphabetical.

Does this help with your use case?

**Comment by mareklibra** [comment]
It would definitely make the "plugins", like these "providers", covered by more general framework which is consistently reused by other packages.
Moreover, the "provider chaining" might be more easily achieved.

So yes, it will be beneficial, no matter the loading is short and simple within recent implementation.

Do I understand correctly, your effort is WIP?

**Comment by stefwalter** [comment]
> Do I understand correctly, your effort is WIP?

Yes, and it shouldn't provide a blocker to this pull request ... but when it's in place (soon) I'd like to have the  machines code updated to use it.

File: pkg/machines/helpers.es6 [hunk 1]
```diff
@@ -87,14 +90,24 @@ export function logDebug(msg) {
...
-export function rephraseUI(key, original) {
-    const transform = {
-        'autostart': {
-            'disable': 'disabled',
-            'enable': 'enabled'
-        }
-    };
+const transform = {
+    'autostart': {
+        'disable': _('disabled'),
```
Comment (stefwalter) [inline pkg/machines/helpers.es6]: We use the gettext C parser, so translatable strings need to have double quotes.

File: pkg/machines/index.es6 [hunk 2]
```diff
@@ -23,7 +23,7 @@ import App from './app.jsx';
 import { getAllVms } from './actions.es6';
 import { logDebug } from './helpers.es6';
 
-function render() {
+function render () {
```
Comment (stefwalter) [inline pkg/machines/index.es6]: Unnecessary change. In particular in our code style we usually have functions without a space after their name.

File: pkg/machines/index.es6 [hunk 3]
```diff
@@ -33,7 +33,7 @@ function render() {
 /**
  * Start the application.
  */
-export function appMain() {
+export function appMain () {
```
Comment (stefwalter) [inline pkg/machines/index.es6]: Ditto.

File: pkg/machines/index.html [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
   <script type="text/javascript" src="../base1/cockpit.js"></script>
   <script type="text/javascript" src="../base1/jquery.js"></script>
-  <script type="text/javascript" src="d3.js"></script>
+<!--  <script type="text/javascript" src="d3.js"></script> -->
```
Comment (stefwalter) [inline pkg/machines/index.html]: Lets remove this rather than comment it out. Perhaps a separate commit?

**Comment by stefwalter** [comment]
Two other questions:

 * Can multiple providers be used at once and contribute VMs to the displayed list?
 * Does this pull request change anything visible in the UI?

**Comment by mareklibra** [comment]
| - Can multiple providers be used at once and contribute VMs to the displayed list?

Recently just the default Libvirt provider is chained and used as a fallback if external provider's method implementation is missing. It other words, the dynamically loaded external provider can redirect/reuse it's actions/properties. 

If real chaining is required in the future, I would extend provider.es6:loadExternalProvider() to implement it.

For completeness, internal data structure for a VM keeps information about it's origin so it can be used to match corresponding external provider in such a chain.



| - Does this pull request change anything visible in the UI?

Localization of VM states only.

File: pkg/machines/index.html (hunk 4 above)
Comment (mareklibra) [inline pkg/machines/index.html]: Done

File: pkg/machines/index.es6 (hunk 3 above)
Comment (mareklibra) [inline pkg/machines/index.es6]: Done

File: pkg/machines/helpers.es6 (hunk 1 above)
Comment (mareklibra) [inline pkg/machines/helpers.es6]: Done

**Comment by mareklibra** [comment]
rebased

**Comment by mareklibra** [comment]
With last commit, the React virtual DOM can be extended for components created by an external provider.
With this feature, we might (almost) fully plug functionality into existing code base.

**Comment by stefwalter** [comment]
@mareklibra Could you rebase this pull request?

**Comment by mareklibra** [comment]
Rebased

**Comment by mareklibra** [comment]
Failing tests are unrelated

**Comment by stefwalter** [comment]
@mareklibra Would you be able to look into the test failure in check-machines on fedora-25 above?

```
# testBasic (check_machines.TestMachines)
#
Ignoring Resource Error: Operation canceled http://10.111.118.33:9090/cockpit/$908792905a1c5bfb33554d710171359197bd8e48/machines/provider/index.js
WARNING  The guest's network configuration does not support PXE
WARNING  No operating system detected, VM performance may suffer. Specify an OS with --os-variant for optimal results.
Wrote TestMachines-testBasic-FAIL.png
Journal extracted to TestMachines-testBasic-10.111.118.33-FAIL.log
not ok 14 testBasic (check_machines.TestMachines) duration: 98s
Traceback (most recent call last):
  File "./verify/check-machines", line 59, in testBasic
    b.wait_in_text("{0}-state".format(vm1_id), "running")
  File "/build/cockpit/test/common/testlib.py", line 242, in wait_in_text
    return self.wait_js_func('ph_in_text', selector, text)
  File "/build/cockpit/test/common/testlib.py", line 209, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 722, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 745, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by mareklibra** [comment]
This test failure seems to be the same as in https://github.com/cockpit-project/cockpit/pull/5835#issuecomment-278144551 .

**Comment by stefwalter** [comment]
Lets try https://github.com/cockpit-project/cockpit/pull/5866

**Comment by mareklibra** [comment]
Rebased

**Comment by mareklibra** [comment]
Failing tests are unrelated
